### PR TITLE
Fix/race condition

### LINKED ISSUE #39: Race condition
tbd

### PR DESCRIPTION
**Pull request by Dexus**
I need at last three reviews, thank you.

To merge and 
fix #39

### CODE REVIEW

**Review by theodesp — APPROVED** [review]
Looks good

**Review by theodesp — APPROVED** [review]
Looks good. Simpler than before

**Review by theodesp — APPROVED** [review]
Looks good

**Review by theodesp — APPROVED** [review]
Amazing

File: core/host/supervisor.go [hunk 1]
```diff
@@ -258,10 +226,33 @@ func (su *Supervisor) ListenAndServeAutoTLS() error {
...
+	su.manuallyTLS = true
 	return su.ListenAndServe()
 }
 
+// RegisterOnShutdown registers a function to call on Shutdown.
+// This can be used to gracefully shutdown connections that have
+// undergone NPN/ALPN protocol upgrade or that have been hijacked.
+// This function should start protocol-specific graceful shutdown,
+// but should not wait for shutdown to complete.
+func (su *Supervisor) RegisterOnShutdown(cb func()) {
```
Comment (theodesp) [inline core/host/supervisor.go]: Maybe a better name for this could be `RegisterShutdownHook`

File: core/errors/errors.go [hunk 2]
```diff
@@ -46,34 +68,53 @@ func (e Error) Error() string {
 // Format returns a formatted new error based on the arguments
 // it does NOT change the original error's message
 func (e Error) Format(a ...interface{}) Error {
-	e.message = fmt.Sprintf(e.message, a...)
+	e.Message = fmt.Sprintf(e.Message, a...)
+	return e
+}
+
+func omitNewLine(message string) string {
+	if strings.HasSuffix(message, "\n") {
```
Comment (godofdream) [inline core/errors/errors.go]: I think \\n should be checked first and \n afterwards

**Comment by godofdream** [comment]
Why do you replace the logger? logrus is nice but definitly not the fastest logger. I'm using e.g. zap.

**Comment by Dexus** [comment]
I have took the changes from Iris/Ion.
I did not know zap, but that would be a very good alternative.

**Comment by godofdream** [comment]
logrus is nice but definitly not fast,
see some benchmarks: https://github.com/uber-go/zap

**Review by Andersbiha — APPROVED** [review]
LGTM.
I would definatly put my vote on replacing Logrus with Zap at some point.
The performance gain have turned out to be massive in other projects of mine.

**Comment by Dexus** [comment]
Okay, then I'll remove logrus and add Zap.

Any hints for zap, as i use it the first time?

File: core/host/supervisor.go (hunk 1 above)
Comment (Dexus) [reply core/host/supervisor.go]: @go-siris/core any thoughts from others?

File: core/errors/errors.go (hunk 2 above)
Comment (Dexus) [reply core/errors/errors.go]: Have now changed it. Hope I'm right with this integration like you mean it should be done.

File: core/host/supervisor.go (hunk 1 above)
Comment (Dexus) [reply core/host/supervisor.go]: Have changed the name to `RegisterOn...Hook`

**Comment by Dexus** [comment]
@godofdream can you now again review?
Zap is added and the "newline" is also added... but it should never used only when created a new error from `/core/errors` with any `Append` function on it.

**Review by godofdream — APPROVED** [review]
Seems fine. I tried it in some of my projects and it seems fine.

**Pull request merged**